waiter/src/waiter/auth/saml.clj

```diff
@@ -0,0 +1,381 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns waiter.auth.saml
+  (:require [clj-time.coerce :as tc]
+            [clj-time.core :as t]
+            [clj-time.format :as f]
+            [clojure.data.codec.base64 :as b64]
+            [clojure.java.io :as io]
+            [clojure.string :as string]
+            [comb.template :as template]
+            [metrics.counters :as counters]
+            [plumbing.core :as pc]
+            [ring.middleware.params :as ring-params]
+            [ring.util.codec :as codec]
+            [ring.util.response :as response]
+            [waiter.auth.authentication :as auth]
+            [waiter.middleware :as middleware]
+            [waiter.util.date-utils :as du]
+            [waiter.util.utils :as utils]
+            [waiter.metrics :as metrics])
+  (:import (java.io ByteArrayInputStream ByteArrayOutputStream)
+           (java.nio.charset Charset)
+           (java.security.cert CertificateFactory)
+           (java.util.zip Deflater DeflaterOutputStream)
+           (javax.xml.parsers DocumentBuilderFactory)
+           (org.opensaml Configuration DefaultBootstrap)
+           (org.opensaml.xml.security.x509 BasicX509Credential)
+           (org.opensaml.xml.signature SignatureValidator)
+           (org.opensaml.xml.validation ValidationException)))
+
+(def charset-format (Charset/forName "UTF-8"))
+
+(defn str->bytes
+  "Get bytes of a string"
+  [some-string]
+  (.getBytes some-string charset-format))
+
+(defn bytes->str
+  "Get string from bytes"
+  [some-bytes]
+  (String. some-bytes charset-format))
+
+(defn deflate-bytes
+  "Gzip compress bytes"
+  [str-bytes]
+  (let [out (ByteArrayOutputStream.)
+        deflater (DeflaterOutputStream.
+                   out
+                   (Deflater. -1 true) 1024)]
+    (.write deflater str-bytes)
+    (.close deflater)
+    (.toByteArray out)))
+
+(defn str->deflate->base64
```

Comment:
I read `a->b->c`as it will give me a `b->c` if I give an `a` as input.

I think this function is better named as `deflate-and-base64-encode`